### ISSUE TITLE
no attention backend support for softmax = learnable and cp_comm_type = p2p

### ISSUE THREAD

**Issue by jordane95** [NVIDIA/TransformerEngine#2541]
When using context parallel to fine-tune gpt-oss, no attention backend is supported for this configuration. I have to change cp_comm_type to a2a to enable FusedAttention. But this is potentially less efficient than a2a at large context length (say 128k).

**Comment by 0z5a** [comment]
I reproduced the reported combination on L20 and have the capability matrix plus a sink-math oracle. Short version: P2P context parallelism itself is healthy here, only the learnable cell is blocked, and the block is a parameter-plumbing gap on the p2p fused path rather than a kernel problem.

**Environment**: 2× L20 (cc 8.9), TE `main` @ `9233c73a`, torch 2.11.0+cu128, cuDNN 9.19, NCCL 2.28.9, `NVTE_ALLOW_NONDETERMINISTIC_ALGO=1`. The CP communicator must be installed with `set_context_parallel_group(cp_group, cp_ranks, cp_stream, cp_comm_type=...)`, as `tests/pytorch/attention/run_attention_with_cp.py` does. Note that passing only the `DotProductAttention(cp_group=...)` constructor argument is not sufficient — the CP runtime needs the stream too, and without it every CP cell fails even for vanilla softmax.

**1. Capability matrix** (`b=2, s=4096, heads=32, gqa=4, head_dim=128`, bf16, `attn_mask_type="no_mask"`)

| cp_comm_type | softmax_type | forward | backward | sink grad |
|---|---|---|---|---|
| no_cp | vanilla | ok | ok | — |
| no_cp | learnable | ok | ok | present, non-zero |
| **p2p** | vanilla | **ok** | **ok** | — |
| **p2p** | **learnable** | **refused** | — | — |
| a2a | vanilla | ok | ok | — |
| a2a | learnable | ok | ok | present, non-zero |

The p2p/learnable refusal is:

```
No dot product attention backend is available for the provided inputs.
```

So p2p CP itself works on this machine (vanilla p2p runs forward and backward), and a2a supports learnable end to end. The failure is specific to the combination.

**2. Where it is blocked**

`attention/dot_product_attention/utils.py`:

```python
if context_parallel and cp_comm_type != "a2a":
    logger.debug("Disabling FusedAttention for context parallelism with softmax_type = %s"
                 " cp_comm_type = %s", softmax_type, cp_comm_type)
    use_fused_attention = False
```

and `context_parallel.py` only has the offset-communication helper for the a2a path:

```
flash_attn_a2a_communicate_softmax_offset   # split round-robin to CP ranks / all-gather on the way back
```

The p2p fused functions take no `softmax_offset` at all — `cp_p2p_fwd_fused_attn` has zero mentions of it, while it does already thread `softmax_lse` through its aux tensors (`softmax_lse_per_step = aux_ctx_tensors[0]`, and the matching backward takes `softmax_lse, softmax_lse_`). So the fused kernel is already being handed per-step normalisation statistics; what is missing on that path is the offset parameter and the sink-gradient return.

**The kernel contract itself is fine.** On the no-CP path (where learnable softmax works) the fused backend already accepts the offset and hands back what a sink-aware backward needs:

| configuration | output | extra aux | softmax_offset | dsink |
|---|---|---|---|---|
| vanilla, `return_max_logit=False` | `Tensor [b,s,h,d]` | — | none | — |
| vanilla, `return_max_logit=True` | tuple `(out, max_logit)` | `[heads]` | none | — |
| learnable, `return_max_logit=False` | `Tensor [b,s,h,d]` | — | `[heads]` | present, non-zero |
| learnable, `return_max_logit=True` | tuple `(out, max_logit)` | `[heads]` | `[heads]` | present, non-zero |

So on this backend the offset parameter goes in and a usable sink gradient comes out; the p2p CP wrapper is the only piece not wired for it.

**3. Sink math oracle** (independent FP64 implementation, single GPU)

The convention in `common/include/transformer_engine/fused_attn.h` is `S = exp(S) / (exp(alpha) + sum(exp(S)))`, i.e. one logit `s = alpha[j]` per query/head whose value is zero, and the whole-query denominator contains the sink **once**:

```
L_real = logsumexp_r(l_r);  L = logaddexp(L_real, s)
O      = sum_r exp(l_r - L) * O_r;  p_sink = exp(s - L)
dlogit_i = p_i * (dot(g, V_i) - dot(g, O_global));  ds = -p_sink * dot(g, O_global)
```

Verification:

* row probability mass **plus** sink probability = **1.000000** for sink values 0.0 / 0.5 / 8.0 (sink probability 0.043 / 0.069 / 0.991)
* analytic `ds` vs central finite differences: max abs diff **1.09e-09**
* dQ/dK/dV finite

Two things this pins down for anyone implementing it: the backward needs **`O_global`**, not the local partition's output, and any implementation that adds the same sink to every partition before summing will double-count it.

**4. What I have and have not done**

Done: the matrix above, the oracle, the localisation to the p2p fused parameter chain, and a work-in-progress patch that threads `softmax_offset` through `cp_p2p_fwd_fused_attn` / `cp_p2p_bwd_fused_attn`, saves it on the autograd context, accumulates the per-tile sink gradient, and relaxes the selector clause for p2p. That patch now gets p2p past the fused-forward argument binding (an earlier argument-position mistake on my side made the offset land in the wrong slot); no_cp and a2a are unaffected and still pass. It does not yet run end to end, so I am not claiming support.

Not done: the runtime change is not complete and I have not opened a PR.

**5. What I would like to know**

1. Is the intended fix to thread `softmax_offset` through `cp_p2p_fwd_fused_attn` / `cp_p2p_bwd_fused_attn` (and lift the `cp_comm_type != "a2a"` clause only for the learnable case), or is there a reason the sink is a2a-only that I am missing?
2. For p2p, does the sink gradient need the same cross-rank treatment as a2a? My reading is that each rank owns whole query heads so the per-head sink contribution is rank-local and only the replicated parameter needs the final reduction, but the backward still needs the global `O`/`L`, which is where I would expect the real work to be. I would rather confirm than guess.
3. `softmax_offset` shape: the a2a helper splits it across ranks round-robin; for p2p the natural thing looks like keeping the full per-head vector on every rank. Does that match your intent?

**Comment by 0z5a** [comment]
I reproduced the reported combination on L20 and have the capability matrix plus a sink-math oracle. Short version: the block is exactly where the source says it is, and the fix is a parameter-plumbing job on the p2p fused path rather than a kernel problem.

**Environment**: 2× L20 (cc 8.9), TE `main` @ `9233c73a`, torch 2.11.0+cu128, cuDNN 9.19, NCCL 2.28.9, `NVTE_ALLOW_NONDETERMINISTIC_ALGO=1`. CP communicator installed via `set_context_parallel_group(cp_group, cp_ranks, cp_stream, cp_comm_type=...)` as `tests/pytorch/attention/run_attention_with_cp.py` does — note that passing only the `DotProductAttention(cp_group=...)` constructor argument is not sufficient, the CP runtime needs the stream too.

**1. Capability matrix** (`b=2, s=4096, heads=32, gqa=4, head_dim=128`, bf16, `attn_mask_type="no_mask"`)

| cp_comm_type | softmax_type | forward | backward | sink grad |
|---|---|---|---|---|
| no_cp | vanilla | ok | ok | — |
| no_cp | learnable | ok | ok | present, non-zero |
| **p2p** | vanilla | **ok** | **ok** | — |
| **p2p** | **learnable** | **refused** | — | — |
| a2a | vanilla | ok | ok | — |
| a2a | learnable | ok | ok | present, non-zero |

The p2p/learnable refusal is:

```
No dot product attention backend is available for the provided inputs.
```

So p2p context parallelism itself is healthy on this machine (vanilla p2p runs forward and backward), and the a2a setup does support learnable softmax end to end. The failure is specific to the combination.

**2. Where it is blocked**

`attention/dot_product_attention/utils.py`:

```python
if context_parallel and cp_comm_type != "a2a":
    logger.debug("Disabling FusedAttention for context parallelism with softmax_type = %s"
                 " cp_comm_type = %s", softmax_type, cp_comm_type)
    use_fused_attention = False
```

and `context_parallel.py` only has the offset-communication helper for the a2a path:

```
flash_attn_a2a_communicate_softmax_offset   # split round-robin to CP ranks / all-gather on the way back
```

The p2p fused functions take no `softmax_offset` at all — `cp_p2p_fwd_fused_attn` has zero mentions of it, while it does already thread `softmax_lse` through its aux tensors (`softmax_lse_per_step = aux_ctx_tensors[0]`, and the matching backward takes `softmax_lse, softmax_lse_`). So the fused kernel is already being handed per-step normalisation statistics; what is missing on that path is the offset parameter and the sink-gradient return.

**The kernel contract itself is fine.** On the no-CP path (where learnable softmax works) the fused backend already accepts the offset and hands back what a sink-aware backward needs:

| configuration | output | extra aux | softmax_offset | dsink |
|---|---|---|---|---|
| vanilla, `return_max_logit=False` | `Tensor [b,s,h,d]` | — | none | — |
| vanilla, `return_max_logit=True` | tuple `(out, max_logit)` | `[heads]` | none | — |
| learnable, `return_max_logit=False` | `Tensor [b,s,h,d]` | — | `[heads]` | present, non-zero |
| learnable, `return_max_logit=True` | tuple `(out, max_logit)` | `[heads]` | `[heads]` | present, non-zero |

So on this backend the offset parameter goes in and a usable sink gradient comes out; the p2p CP wrapper is the only piece not wired for it.

**3. Sink math oracle** (independent FP64 implementation, single GPU)

The convention in `common/include/transformer_engine/fused_attn.h` is `S = exp(S) / (exp(alpha) + sum(exp(S)))`, i.e. one logit `s = alpha[j]` per query/head whose value is zero, and the whole-query denominator contains the sink **once**:

```
L_real = logsumexp_r(l_r);  L = logaddexp(L_real, s)
O      = sum_r exp(l_r - L) * O_r;  p_sink = exp(s - L)
dlogit_i = p_i * (dot(g, V_i) - dot(g, O_global));  ds = -p_sink * dot(g, O_global)
```

Verification:

* row probability mass **plus** sink probability = **1.000000** for sink values 0.0 / 0.5 / 8.0 (sink probability 0.043 / 0.069 / 0.991)
* analytic `ds` vs central finite differences: max abs diff **1.09e-09**
* dQ/dK/dV finite

Two things this pins down for anyone implementing it: the backward needs **`O_global`**, not the local partition's output, and any implementation that adds the same sink to every partition before summing will double-count it.

**4. What I have and have not done**

Done: the matrix above, the oracle, and the localisation to the p2p fused parameter chain.

Not done: the runtime change itself. I have not touched `context_parallel.py`, and I am not claiming p2p support works. Before I write it I would like to know:

1. Is the intended fix to thread `softmax_offset` through `cp_p2p_fwd_fused_attn` / `cp_p2p_bwd_fused_attn` (and lift the `cp_comm_type != "a2a"` clause only for the learnable case), or is there a reason the sink is a2a-only that I am missing?
2. For p2p, does the sink gradient need the same cross-rank treatment as a2a? My reading is that each rank owns whole query heads so the per-head sink contribution is rank-local and only the replicated parameter needs the final reduction, but the backward still needs the global `O`/`L`, which is where I would expect the real work to be. I would rather confirm than guess.
3. `softmax_offset` shape: the a2a helper splits it across ranks round-robin; for p2p the natural thing looks like keeping the full per-head vector on every rank. Does that match your intent?

Happy to take the runtime piece once the shape of it is settled — I just do not want to send a PR that silently redefines the semantics.

**Comment by 0z5a** [comment]
Root-caused this and opened a draft PR: https://github.com/NVIDIA/TransformerEngine/pull/3551

The p2p fused-attention helpers never passed `softmax_type` to `fused_attn_fwd` / `fused_attn_bwd`, so the kernels used their `"vanilla"` default and silently discarded `softmax_offset`. That is why the assertion was the visible symptom rather than the whole problem — even with the assertion relaxed, the sink was a no-op: the forward returned the vanilla result and the sink gradient came back as an exact zero.

One trap worth recording, because it cost me a wrong conclusion first: a sink column with logit `0` contributes `exp(-max_logit)` to a denominator of order `Z`, a relative change near `1e-6`, which BF16 rounds away. Comparing vanilla against learnable with the default zero-initialised offset therefore shows *bitwise identical* output even when the sink is being applied correctly. I initially read that as "sink not applied". Driving the offset to `20` is what made it a real, reproducible measurement.

After threading `softmax_type` through, against an FP64 oracle (autograd and finite differences agree to `2.3e-08`) at cp=2, BF16, 1024 tokens:

| softmax | comm | out rel err | dSink rel err |
|---|---|---|---|
| learnable | a2a | 9.494e-03 | 1.370e-02 |
| learnable | p2p | 9.494e-03 | 1.696e-03 |

Not covered by the PR: `qkv_format="thd"`, cp_size > 2, and the flash-attention backend (not installable on my test box, so that branch is guarded but unexercised).

**Comment by 0z5a** [comment]
#3551 is out of draft and ready for review.

**Comment by sudhakarsingh27** [comment]
Thanks! Will take a look but I'm curious if you have perf numbers to support this: 
> But this is potentially less efficient than a2a at large context length (say 128k).
